chalk-integration/src/program.rs

```diff
@@ -100,21 +106,46 @@ impl tls::DebugContext for Program {
         assoc_type_id: AssocTypeId<ChalkIr>,
         fmt: &mut fmt::Formatter<'_>,
     ) -> Result<(), fmt::Error> {
-        if let Some(k) = self.associated_ty_data.get(&assoc_type_id) {
-            write!(fmt, "({:?}::{})", k.trait_id, k.name)
+        if let Some(d) = self.associated_ty_data.get(&assoc_type_id) {
+            write!(fmt, "({:?}::{})", d.trait_id, d.name)
         } else {
             fmt.debug_struct("InvalidItemId")
                 .field("index", &assoc_type_id.0)
                 .finish()
         }
     }
 
+    fn debug_opaque_ty_id(
+        &self,
+        opaque_ty_id: OpaqueTyId<ChalkIr>,
+        fmt: &mut fmt::Formatter<'_>,
+    ) -> Result<(), fmt::Error> {
+        if let Some(d) = self.opaque_ty_data.get(&opaque_ty_id) {
+            write!(fmt, "{:?}", d.bound.skip_binders().hidden_ty)
```

Comment:
...or maybe here we want to print the name? anyway, somewhere. It'd be nice for opaque types to print out with their name and for the placeholders to print with `!Foo`.